server/src/main/java/au/org/aodn/ogcapi/server/core/service/ElasticSearch.java

```diff
@@ -215,74 +221,153 @@ protected List<Hit<CategorySuggestDTO>> getCategorySuggestions(String input) thr
+     */
+    protected <T> Iterable<T> pagableSearch(Supplier<SearchRequest.Builder> requestBuilder, Class<T> clazz, Long maxSize) {
+        try {
+            SearchRequest sr = requestBuilder.get().build();
+            logger.debug("Final elastic search payload {}", sr.toString());
 
-        SearchRequest request = builder.build();
-        logger.debug("Final elastic search payload {}", request.toString());
+            final AtomicLong count = new AtomicLong(0);
+            final AtomicReference<SearchResponse<T>> response = new AtomicReference<>(
+                    esClient.search(sr, clazz)
+            );
 
-        try {
-            SearchResponse<ObjectNode> response = esClient.search(request, ObjectNode.class);
-
-            return response
-                    .hits()
-                    .hits()
-                    .stream()
-                    .map(this::formatResult)
-                    .filter(Objects::nonNull)
-                    .collect(Collectors.toList());
+            return () -> new Iterator<>() {
+                private int index = 0;
+
+                @Override
+                public boolean hasNext() {
+                    // If we hit the end, that means we have iterated to end of page.
+                    if (index < response.get().hits().hits().size()) {
+                        if(maxSize != null) {
+                            return count.get() < maxSize;
+                        }
+                        else {
+                            return true;
+                        }
+                    }
+                    else {
+                        // If last index is zero that mean nothing found already, so no need to look more
+                        if (index == 0) return false;
+
+                        // Load next batch
+                        try {
+                            // Get the last sorted value from the last batch
+                            List<FieldValue> sortedValues = response.get().hits().hits().get(index - 1).sort();
+
+                            // Use the last builder and append the searchAfter values
+                            SearchRequest request = requestBuilder.get().searchAfter(sortedValues).build();
+                            logger.debug("Final elastic search payload {}", request.toString());
+
+                            response.set(esClient.search(request, clazz));
+                            // Reset counter from start
+                            index = 0;
+                            return index < response.get().hits().hits().size();
+                        }
+                        catch(IOException ieo) {
+                            throw new RuntimeException(ieo);
+                        }
+                    }
+                }
+
+                @Override
+                public T next() {
+                    count.incrementAndGet();
+                    return response.get().hits().hits().get(index++).source();
```

Comment:
how is this related to the PR yesterday that you thought it was a typo? https://github.com/aodn/es-indexer/pull/87

Reply:
No this one is fine because the index++ is apply after the .get(), yesterday PR is fix because the index is pass in as a value, hence there is no point to do ++, after the call.